firmware/library/L1_Drivers/pwm.hpp

```diff
@@ -1,13 +1,176 @@
+    // The variable that holds the MR values
+    static constexpr volatile uint32_t * match_register_table[7] = 
+    {
+        & LPC_PWM1->MR0,
+        & LPC_PWM1->MR1,
+        & LPC_PWM1->MR2,
+        & LPC_PWM1->MR3,
+        & LPC_PWM1->MR4,
+        & LPC_PWM1->MR5,
+        & LPC_PWM1->MR6
+    };
+
+
+    template <unsigned pwm_port, unsigned pwm_pin>
+    static Pwm CreatePwm()
+    {
+        static_assert(1 <= pwm_port && pwm_port <= 2,
+                      "Available PWM ports are only on port 1 or 2.");
+        static_assert(pwm_pin <= 5,
+                      "Available PWM pins only go up to pin 5.");
+        return Pwm(pwm_port, pwm_pin);
+    }
+
+    constexpr Pwm(uint8_t vpport, uint8_t vppin):
+        pwm_(pwm_pin_),
+        pwm_pin_(vpport, vppin),
+        port_(vpport),
+        channel_(vppin),
+        duty_cycle(0)
+    {
+        // Sets port to proper function and get pwm channel
+        if (pwm_pin_.GetPort() == 2)
+        {
+            port_ = 1;
+            channel_ = static_cast<uint8_t>(pwm_pin_.GetPin() + 1);
+        }
+    }
+
+    Pwm(const PinConfigureInterface & gpio_):
+        pwm_(gpio_),
+        pwm_pin_(PinConfigure::CreateInactivePin()),
+        port_(pwm_pin_.GetPort()),
+        channel_(pwm_pin_.GetPin()),
+        duty_cycle(0)
+    {
+    }
+
+
+    ~Pwm()
+    {
+        pwm_pin_.SetPinFunction(kGpioFunction);
+    }
+
+    void Initialize(uint32_t frequency_hz = 1'000) override
+    {
+        // TODO: Remove when sysclock complete.
+        // DIRECT MANIUPULATION OF SYSTEM CLOCK REGISTERS
+        // WILL REPLACE WITH FUNCTIONS LATER
+        // Enables PWM1 power/clock control bit
+        LPC_SC->PCONP |= SysClk::kActivatePwm1;
+        // END OF DIRECT MANIPULATION
+
+        match_value = config::kSystemClockRate / frequency_hz;
+
+        // Resets PWMTC on Match with MR0
+        LPC_PWM1->MCR |= PwmConfigure::kResetMr0;
+        LPC_PWM1->MR0 = match_value;
+
+        // Enables PWM TC and PC for counting and enables PWM mode
+        LPC_PWM1->TCR = PwmConfigure::kCounterEnable | PwmConfigure::kPwmEnable;
+        LPC_PWM1->CTCR &= PwmConfigure::kTimerMode | PwmConfigure::kUsePwmCap0;
+
+
+        // Enables PWM[channel] output
+        LPC_PWM1->PCR |= PwmOutputEnable(channel_);
+
+        pwm_pin_.SetPinFunction(kPwmFunction);
+    }
+
+    void SetDutyCycle(float percentage) override
+    {
+
+        if (percentage < 0 || percentage > 100)
+        {
+            // TODO: Replace with an assert 
+            return;
+        }
+        duty_cycle = static_cast<uint32_t>((percentage * static_cast<float>(match_value)) / 100);
+
+        // Pointer to point to address of Match Register
+        volatile uint32_t * match_register_selector;
+
+        // Have the pointer point to address of MR
+        match_register_selector = match_register_table[channel_];
+        // Set the value of the MR
+        *match_register_selector = duty_cycle;
```

Comment:
I'm not sure why you went to the lengths to make a pointer to the register when you can use it directly. It would look cleaner if you simply use the table. Changing the table from static to constexpr should fix any compiler errors you may have gotten with this.